data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/HashRing.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.peerforwarder;
+
+import com.amazon.dataprepper.peerforwarder.discovery.PeerListProvider;
+import com.linecorp.armeria.client.Endpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+
+/**
+ * Consistent hashing implementation used to map identification keys to Data Prepper hosts.
+ * See https://en.wikipedia.org/wiki/Consistent_hashing for more information.
+ */
+@NotThreadSafe
+public class HashRing implements Consumer<List<Endpoint>> {
+    private static final Logger LOG = LoggerFactory.getLogger(HashRing.class);
+    private static final String MD5 = "MD5";
+    private static final String DELIMITER = ",";
+
+    /* Number of virtual nodes per Data Prepper host to be present on the hash ring */
+    private final int numVirtualNodes;
+
+    private final PeerListProvider peerListProvider;
+
+    private TreeMap<BigInteger, String> hashServerMap = new TreeMap<>();
+
+    public HashRing(final PeerListProvider peerListProvider, final int numVirtualNodes) {
+        Objects.requireNonNull(peerListProvider);
+        this.peerListProvider = peerListProvider;
+        this.numVirtualNodes = numVirtualNodes;
+
+        buildHashServerMap();
+
+        peerListProvider.addListener(this);
+    }
+
+    public Optional<String> getServerIp(final Set<String> identificationKeys) {
+        if (hashServerMap.isEmpty()) {
+            return Optional.empty();
+        }
+
+        final byte[] identificationKeysInBytes = String.join(DELIMITER, identificationKeys).getBytes();
```

Comment:
It would be good to inject this. Not so much for customization, but to facilitate separation of concerns and more thorough testing. It's not necessary to resolve this PR though.